R/tidyeval.R

```diff
@@ -351,3 +359,10 @@ check_one_arg <- function(x) {
     abort(glue("`{fun}()` expects exactly one argument."))
   }
 }
+
+prep_case_match_dot <- function(dot, .x) {
+  lhs <- f_lhs(dot)
+  lhs <- call2("%in%", .x, lhs)
```

Comment:
Would it make sense to have it as below, to use `==` when a literal is provided?
```r
  if (is.numeric(lhs) || is.character(lhs)) {
    lhs <- call2("==", .x, lhs)
  } else {
    lhs <- call2("%in%", .x, lhs)
  }
```

It does seem to make a difference in speed. Not sure if I'm missing any edge cases by adding this though.

``` r
suppressPackageStartupMessages({
library(data.table)
library(dplyr)
library(bench)
})

b_case_match <- function(x) {
  case_match(x, 
             'a' ~ 1, 
             'b' ~ 2, 
             'c' ~ 3)
}
b_fcase <- function(x) {
  fcase(x %in% 'a', 1, 
        x %in% 'b', 2, 
        x %in% 'c', 3)
}
b_fcase_eq <- function(x) {
  fcase(x == 'a', 1, 
        x == 'b', 2, 
        x == 'c', 3)
}

#x <- sample(c(letters, LETTERS, paste0(letters, LETTERS)), 1e6, TRUE)
x <- sample(letters, 1e6, TRUE)

mark(b_case_match(x), b_fcase(x), b_fcase_eq(x))
#> Warning: Some expressions had a GC in every iteration; so filtering is disabled.
#> # A tibble: 3 × 6
#>   expression           min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>      <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 b_case_match(x)   96.1ms    150ms      7.06    72.8MB     17.6
#> 2 b_fcase(x)       111.5ms    114ms      8.75    57.2MB     15.8
#> 3 b_fcase_eq(x)     29.9ms     33ms     24.7     22.9MB     11.4
```

<sup>Created on 2023-04-17 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

Reply:
I expanded the test a little bit to include longer strings (4 characters), `data.table::%chin%`, and `vctrs::vec_in()`. (I realize this test would detect a symbol and not a literal but for the sake of the test I stored the strings in a variable.)

It seems to me the best option would be:
* If it's a literal use `==`
* If it's a symbol use `vctrs::vec_in()`

Is using `vctrs::vec_in()` against the spirit of `dtplyr` though? We use `rlang::seq2()` in [slice](https://github.com/tidyverse/dtplyr/blob/main/R/step-subset-slice.R#L76).

``` r
suppressPackageStartupMessages({
  library(data.table)
  library(dplyr)
  library(bench)
  library(vctrs)
})

set.seed(1)
strings <- stringi::stri_rand_strings(1000, 4)

str_sample <- sample(strings, 3)
str1 <- str_sample[1]
str2 <- str_sample[2]
str3 <- str_sample[3]

b_case_match <- function(x) {
  case_match(x, 
             str1 ~ 1, 
             str2 ~ 2, 
             str3 ~ 3)
}
b_fcase <- function(x) {
  fcase(x %in% str1, 1, 
        x %in% str2, 2, 
        x %in% str3, 3)
}
b_fcase_eq <- function(x) {
  fcase(x == str1, 1, 
        x == str2, 2, 
        x == str3, 3)
}
b_fcase_chin <- function(x) {
  fcase(x %chin% str1, 1, 
        x %chin% str2, 2, 
        x %chin% str3, 3)
}
b_fcase_vec_in <- function(x) {
  fcase(vec_in(x, str1), 1, 
        vec_in(x, str2), 2, 
        vec_in(x, str3), 3)
}

x <- sample(letters, 1e7, TRUE)

mark(b_case_match(x), b_fcase(x), b_fcase_eq(x), b_fcase_chin(x), b_fcase_vec_in(x),
     iterations = 10)
#> Warning: Some expressions had a GC in every iteration; so filtering is
#> disabled.
#> # A tibble: 5 × 6
#>   expression             min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>        <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 b_case_match(x)      519ms    589ms      1.69     725MB     4.91
#> 2 b_fcase(x)           771ms    816ms      1.18     572MB     1.65
#> 3 b_fcase_eq(x)        218ms    235ms      4.09     229MB     2.46
#> 4 b_fcase_chin(x)      282ms    315ms      3.20     229MB     1.92
#> 5 b_fcase_vec_in(x)    359ms    417ms      2.35     343MB     1.41
```

Reply:
I don't have strong feelings on this one. I think `vec_in()` would be good. Also would probably be fine to leave `case_match` untranslated. There's a line somewhere where we're creating our own custom compilation of various functions to make something faster instead of "simply" translating to a clear data.table equivalent, but using `vec_in()` probably doesn't cross it.